README.md
transmission-fluid
==================

<a href="http://travis-ci.org/#!/edavis/transmission-fluid"><img src="https://secure.travis-ci.org/edavis/transmission-fluid.png" alt="Travis-CI build status"/></a>

A Python wrapper around the [Transmission][transmission] BitTorrent client's [RPC interface][rpc].

```python
>>> from transmission import Transmission
>>> client = Transmission()
>>> response = client('torrent-get', ids=1, fields=['name'])
>>> response['torrents']
[{u'name': u'torrent 1'}]
```

Complete documentation available on [ReadTheDocs][].

[transmission]: http://transmissionbt.com/
[rpc]: https://trac.transmissionbt.com/browser/trunk/extras/rpc-spec.txt
[ReadTheDocs]: http://transmission-fluid.readthedocs.org/


docs/changelog.rst
Changelog
=========

Version 0.7 (TBD)
-----------------

Version 0.6 (April 26, 2014)
----------------------------

* Add Python 3 support

Version 0.4 (September 7, 2013)
-------------------------------

* Bump requests from version 0.13.1 to 1.2.3

Version 0.3 (August 1, 2012)
----------------------------

* Add SSL proxy support
* Drop `anyjson` requirement
* Automatically convert Unix timestamps to :class:`~datetime.datetime` objects

Version 0.2.1 (June 24, 2012)
-----------------------------

* Bugfix: Fix `anyjson` ImportError when loading setup.py

Version 0.2 (June 23, 2012)
---------------------------

* Use setuptools, if available
* Add test suite
* Test against Python 2.6 and 2.7
* Update documentation

Version 0.1 (June 16, 2012)
---------------------------

* Initial release


tests.py
import nose
import mock
import json
import datetime
import unittest
import transmission
from copy import deepcopy

class CopyingMock(mock.MagicMock):
    """
    By default, MagicMock stores args/kwargs by reference rather
    than by value.

    By using deepcopy, we allow ourselves to assert how mocked objects
    were called -- even if a value changes in the mean time.

    http://www.voidspace.org.uk/python/mock/examples.html#coping-with-mutable-arguments
    """
    def __call__(self, *args, **kwargs):
        args = deepcopy(args)
        kwargs = deepcopy(kwargs)
        return super(CopyingMock, self).__call__(*args, **kwargs)

class TestTransmission(unittest.TestCase):
    """
    Base class for Transmission testing.
    """
    def setUp(self):
        self.client = transmission.Transmission()

class TestTransmissionBodyFormatting(TestTransmission):
    """
    Test that requests are formatted properly before being sent to
    Transmission.
    """
    def test_format_body(self):
        """
        Sanity check to make sure requests conform to Transmission API.
        """
        body = self.client._format_request_body(
            "torrent-get", ids=1, fields=["name"])
        expected = {"method": "torrent-get", "tag": 0,
                    "arguments": {"ids": 1, "fields": ["name"]}}
        self.assertEqual(expected, body)

    def test_format_body_replace_underscore(self):
        """
        Underscores get replaced as dashes.
        """
        body = self.client._format_request_body(
            "torrent-set", ids=1, peer_limit=10)
        expected = {"method": "torrent-set", "tag": 0,
                    "arguments": {"ids": 1, "peer-limit": 10}}
        self.assertEqual(expected, body)

class TestTransmissionResponseDeserializing(TestTransmission):
    """
    Test how responses are deserialized and handled.
    """
    @nose.tools.raises(transmission.BadRequest)
    def test_unsuccessful_request(self):
        """
        Any unsuccessful request raises BadRequest.
        """
        with mock.patch("json.loads") as mocked:
            mocked.return_value = {"result": "failure"}
            self.client._deserialize_response(mock.MagicMock())

    @nose.tools.raises(transmission.BadRequest)
    def test_tag_mismatch(self):
        """
        Tag mismatches raises BadRequest.
        """
        with mock.patch("json.loads") as mocked:
            mocked.return_value = {"tag": -1, "result": "success"}
            self.client._deserialize_response(mock.MagicMock())

    def test_tag_incremented_when_matches(self):
        """
        Tag param gets incremented when successful.
        """
        tag = self.client.tag
        with mock.patch("json.loads") as mocked:
            mocked.return_value = {"result": "success", "tag": tag}
            self.client._deserialize_response(mock.MagicMock())
            self.assertEqual(tag + 1, self.client.tag)

    def test_return_arguments_if_exist(self):
        """
        Return the arguments key if it exists.
        """
        with mock.patch("json.loads") as mocked:
            mocked.return_value = {"arguments": [1], "result": "success", "tag": 0}
            ret = self.client._deserialize_response(mock.MagicMock())
            self.assertEqual([1], ret)

    def test_return_none_when_no_arguments(self):
        """
        Return "None" if arguments doesn't exist or is False.
        """
        # "arguments" exists but is False
        with mock.patch("json.loads") as mocked:
            mocked.return_value = {"arguments": [], "result": "success", "tag": self.client.tag}
            ret = self.client._deserialize_response(mock.MagicMock())
            self.assertEqual(None, ret)

        # "arguments" doesn't exist at all
        with mock.patch("json.loads") as mocked:
            mocked.return_value = {"result": "success", "tag": self.client.tag}
            ret = self.client._deserialize_response(mock.MagicMock())
            self.assertEqual(None, ret)

    def test_json_deserialization(self):
        obj = {
            'result': 'success',
            'tag': 0,
            'arguments': {
                'foo': 1,
                'bar': 2,
            },
        }
        response = mock.MagicMock(text=json.dumps(obj))
        self.assertEqual({'foo': 1, 'bar': 2}, self.client._deserialize_response(response))

class TestTransmissionJSON(TestTransmission):
    """
    Test the custom JSON encoder and decoder.
    """
    def test_epochs_to_datetimes(self):
        """
        Convert UNIX timestamps into UTC datetimes.
        """
        returned = json.loads(json.dumps({"addedDate": 1320091283}),
                              cls=transmission.TransmissionJSONDecoder)
        expected = {"addedDate": datetime.datetime(2011, 10, 31, 20, 1, 23)}
        self.assertEqual(expected, returned)

    def test_datetimes_to_epochs(self):
        """
        Convert UTC datetimes into UNIX epochs.
        """
        returned = json.dumps({"addedDate": datetime.datetime(2011, 10, 31, 20, 1, 23)},
                              cls=transmission.TransmissionJSONEncoder)
        expected = '{"addedDate": 1320091283}'
        self.assertEqual(expected, returned)

    def test_dates_to_epochs(self):
        """
        Convert UTC dates to UNIX epochs.
        """
        returned = json.dumps({"addedDate": datetime.date(2011, 10, 31)},
                              cls=transmission.TransmissionJSONEncoder)
        expected = '{"addedDate": 1320019200}'
        self.assertEqual(expected, returned)

class TestTransmissionRequests(TestTransmission):
    def test_csrf_handling(self):
        """
        Make sure the CSRF dance goes smoothly.
        """
        headers = {transmission.CSRF_HEADER: '123'}
        with mock.patch("requests.post", new_callable=CopyingMock) as mocked:
            mocked.side_effect = [mock.MagicMock(name='csrf', status_code=transmission.CSRF_ERROR_CODE, headers=headers),
                                  mock.MagicMock(name='okay')]

            self.client._make_request("torrent-get", ids=1, fields=['name'])
            (csrf, okay) = mocked.mock_calls

            kwargs = csrf[-1]
            self.assertEqual({}, kwargs['headers'])

            kwargs = okay[-1]
            self.assertEqual(headers, kwargs['headers'])


constants.py
TORRENT_GET_FIELDS = [
"activityDate", "addedDate", "bandwidthPriority", "comment", "corruptEver",
"creator", "dateCreated", "desiredAvailable", "doneDate", "downloadDir",
"downloadedEver", "downloadLimit", "downloadLimited", "error", "errorString",
"eta", "files", "fileStats", "hashString", "haveUnchecked", "haveValid",
"honorsSessionLimits", "id", "isFinished", "isPrivate", "isStalled",
"leftUntilDone", "magnetLink", "manualAnnounceTime", "maxConnectedPeers",
"metadataPercentComplete", "name", "peer", "peers", "peersConnected",
"peersFrom", "peersGettingFromUs", "peersSendingToUs", "percentDone",
"pieces", "pieceCount", "pieceSize", "priorities", "queuePosition",
"rateDownload", "rateUpload", "recheckProgress", "secondsDownloading",
"secondsSeeding", "seedIdleLimit", "seedIdleMode", "seedRatioLimit",
"seedRatioMode", "sizeWhenDone", "startDate", "status", "trackers",
"trackerStats", "totalSize", "torrentFile", "uploadedEver",
"uploadLimit", "uploadLimited", "uploadRatio", "wanted", "webseeds",
"webseedsSendingToUs"]


transmission.py
"""
A Python wrapper for Transmission's RPC interface.

>>> from transmission import Transmission
>>> client = Transmission()
>>> client('torrent-get', ids=range(1,11), fields=['name'])
{u'torrents': [
  {u'name': u'Elvis spotted in Florida.mov'},
  {u'name': u'Bigfoot sings the hits'},
  # ...
  {u'name': u'a-song-of-ice-and-fire_final-chapter.txt'}
]}
"""

__version__ = '0.2.1'

import json
import requests

CSRF_ERROR_CODE = 409
CSRF_HEADER = 'X-Transmission-Session-Id'

class BadRequest(Exception): pass

class Transmission(object):
    def __init__(self, host='localhost', port=9091, path='/transmission/rpc',
                 username=None, password=None):
        """
        Initialize the Transmission client.

        The default host, port and path are all set to Transmission's
        default.
        """
        self.url = "http://%s:%d%s" % (host, port, path)
        self.headers = {}
        self.tag = 0

        self.auth = None
        if username or password:
            self.auth = (username, password)

    def __call__(self, method, **kwargs):
        """
        Send request to Transmission's RPC interface.
        """
        response = self._make_request(method, **kwargs)
        return self._deserialize_response(response)

    def _make_request(self, method, **kwargs):
        body = json.dumps(self._format_request_body(method, **kwargs))
        response = requests.post(self.url, data=body, headers=self.headers, auth=self.auth)
        if response.status_code == CSRF_ERROR_CODE:
            self.headers[CSRF_HEADER] = response.headers[CSRF_HEADER]
            return self._make_request(method, **kwargs)
        return response

    def _format_request_body(self, method, **kwargs):
        """
        Create a request object to be serialized and sent to Transmission.
        """
        fixed = {}
        # As Python can't accept dashes in kwargs keys, replace any
        # underscores with them here.
        for k, v in kwargs.iteritems():
            fixed[k.replace('_', '-')] = v
        return {"method": method, "tag": self.tag, "arguments": fixed}

    def _deserialize_response(self, response):
        """
        Return the response generated by the request object, raising
        BadRequest if there were any problems.
        """
        doc = json.loads(response.content)

        if doc['result'] != 'success':
            raise BadRequest("Request failed: '%s'" % doc['result'])

        if doc['tag'] != self.tag:
            raise BadRequest("Tag mismatch: (got %d, expected %d)" % (doc['tag'], self.tag))
        else:
            del doc['tag']
            self.tag += 1

        if 'arguments' in doc:
            return doc['arguments'] or None
        return None


transmission/json_utils.py
"""
A JSON encoder and decoder to simplify working with the RPC's
datatypes.
"""

import json
import calendar
import datetime

# UNIX epochs to be turned into UTC datetimes
TIMESTAMP_KEYS = frozenset(
    ['activityDate',
     'addedDate',
     'dateCreated',
     'doneDate',
     'startDate'])

def epoch_to_datetime(value):
    return datetime.datetime.utcfromtimestamp(value)

def datetime_to_epoch(value):
    if isinstance(value, datetime.datetime):
        return calendar.timegm(value.utctimetuple())
    elif isinstance(value, datetime.date):
        value = datetime.datetime(value.year, value.month, value.day)
        return calendar.timegm(value.utctimetuple())

class TransmissionJSONDecoder(json.JSONDecoder):
    def __init__(self, **kwargs):
        return super(TransmissionJSONDecoder, self).__init__(
            object_hook=self.object_hook, **kwargs)

    def object_hook(self, obj):
        for key, value in obj.items():
            if key in TIMESTAMP_KEYS:
                value = epoch_to_datetime(value)
            obj[key] = value
        return obj

class TransmissionJSONEncoder(json.JSONEncoder):
    def default(self, value):
        # datetime is a subclass of date, so this'll catch both
        if isinstance(value, datetime.date):
            return datetime_to_epoch(value)
        else:
            return value


transmission/__init__.py
"""
A Python wrapper for Transmission's RPC interface.

>>> from transmission import Transmission
>>> client = Transmission()
>>> client('torrent-get', ids=range(1,11), fields=['name'])
{u'torrents': [
  {u'name': u'Elvis spotted in Florida.mov'},
  {u'name': u'Bigfoot sings the hits'},
  # ...
  {u'name': u'a-song-of-ice-and-fire_final-chapter.txt'}
]}
"""

__version__ = '0.7-dev'

import json
import requests
from .json_utils import (
    TransmissionJSONEncoder, TransmissionJSONDecoder)

CSRF_ERROR_CODE = 409
CSRF_HEADER = 'X-Transmission-Session-Id'

class BadRequest(Exception): pass

class Transmission(object):
    def __init__(self, host='localhost', port=9091, path='/transmission/rpc',
                 username=None, password=None, ssl=False):
        """
        Initialize the Transmission client.

        The default host, port and path are all set to Transmission's
        default.
        """
        self.url = "http://%s:%d%s" % (host, port, path)
        if ssl:
            self.url = "https://%s:%d%s" % (host, port, path)
        self.headers = {}
        self.tag = 0

        self.auth = None
        if username or password:
            self.auth = (username, password)

    def __call__(self, method, **kwargs):
        """
        Send request to Transmission's RPC interface.
        """
        response = self._make_request(method, **kwargs)
        return self._deserialize_response(response)

    def _make_request(self, method, **kwargs):
        body = json.dumps(self._format_request_body(method, **kwargs), cls=TransmissionJSONEncoder)
        response = requests.post(self.url, data=body, headers=self.headers, auth=self.auth, verify=False)
        if response.status_code == CSRF_ERROR_CODE:
            self.headers[CSRF_HEADER] = response.headers[CSRF_HEADER]
            return self._make_request(method, **kwargs)
        return response

    def _format_request_body(self, method, **kwargs):
        """
        Create a request object to be serialized and sent to Transmission.
        """
        fixed = {}
        # As Python can't accept dashes in kwargs keys, replace any
        # underscores with them here.
        for k, v in kwargs.items():
            fixed[k.replace('_', '-')] = v
        return {"method": method, "tag": self.tag, "arguments": fixed}

    def _deserialize_response(self, response):
        """
        Return the response generated by the request object, raising
        BadRequest if there were any problems.
        """
        doc = json.loads(response.text, cls=TransmissionJSONDecoder)

        if doc['result'] != 'success':
            raise BadRequest("Request failed: '%s'" % doc['result'])

        if doc['tag'] != self.tag:
            raise BadRequest("Tag mismatch: (got %d, expected %d)" % (doc['tag'], self.tag))
        else:
            self.tag += 1

        if 'arguments' in doc:
            return doc['arguments'] or None
        return None
